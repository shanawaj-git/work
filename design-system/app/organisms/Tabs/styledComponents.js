import { Tab, Tabs } from '@mui/material';
import React from 'react';
import styled, { css } from 'styled-components';
import { colorSelector } from 'themes';

import {
  defaultLabelStyle,
  defaultRootStyle,
  labelStyles,
  rippleStyle,
  tabLabelContainerStyles,
  tabRootStyles,
} from './styles';

export const StyledTab = styled(
  ({ classes, size, label, isActive, ...other }) => (
    <Tab
      {...other}
      label={label}
      TouchRippleProps={{ style: rippleStyle }}
      classes={{
        root: 'root',
      }}
    />
  ),
)`
  &.root {
    min-width: ${props => tabRootStyles[props.size].minWidth};
    padding: ${props => tabLabelContainerStyles[props.size].padding};
    font-family: ${defaultLabelStyle.fontFamily};
    font-weight: ${defaultLabelStyle.fontWeight};
    text-transform: ${defaultLabelStyle.textTransform};

    &:hover {
      color: ${props => labelStyles[`${props.isActive}`].hover.color};
    }
  }

  &&.Mui-selected {
    color: ${colorSelector('White')};
    font-weight: 600;
    font-size: 14px;

    background-color: ${props =>
      labelStyles[`${props.isActive}`].backgroundColor};
    border-radius: 10px;
  }
`;

export const StyledTabs = styled(
  ({
    classes,
    size,
    onChangeCallback,
    preSelectedTab,
    tabs,
    underline,
    ...other
  }) => (
    <Tabs
      {...other}
      classes={{
        root: 'root',
        indicator: 'indicator',
        scrollable: 'scrollable',
        flexContainer: 'flexContainer',
      }}
    />
  ),
)`
  ${props =>
    props.underline &&
    css`
      position: relative;

      ::after {
        content: '';
        position: absolute;
        display: block;
        left: 0;
        bottom: 0;
        right: 0;
        width: 100%;
        border: 0.5px solid ${props.theme.color.g400};
        z-index: ${props.theme.layer.modal};
      }
    `}

  &.root {
    background-color: ${defaultRootStyle.backgroundColor};
    background: ${colorSelector('P100')};
    border-radius: 10px;

    min-height: ${props => tabRootStyles[props.size].minHeight};
  }

  & .indicator {
    visibility: hidden;
  }

  & .scrollable {
    overflow-x: 'scroll';
    scrollbar-width: none;
  }

  & .scrollable::-webkit-scrollbar {
    display: none;
  }
`;
