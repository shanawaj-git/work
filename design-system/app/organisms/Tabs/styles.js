import { FONTS } from 'styledComponents/styles';

import { colorSelector } from 'themes';

export const defaultRootStyle = {
  backgroundColor: 'transparent',
};

export const defaultLabelStyle = {
  fontFamily: 'Poppins',
  fontWeight: '600',
  color: colorSelector('G400'),
  textTransform: 'capitalize',
};

export const defaultTabIndicatorStyle = {
  backgroundColor: colorSelector('P300'),
};

export const sizeStyles = {
  default: {
    fontSize: FONTS.bodyL.fontSize,
    lineHeight: FONTS.bodyL.lineHeight,
  },
  extraSmall: {
    fontSize: FONTS.bodyS.fontSize,
    lineHeight: FONTS.bodyS.lineHeight,
  },
  small: {
    fontSize: FONTS.bodyM.fontSize,
    lineHeight: FONTS.bodyM.lineHeight,
  },
  medium: {
    fontSize: FONTS.bodyL.fontSize,
    lineHeight: FONTS.bodyL.lineHeight,
  },
  large: {
    fontSize: FONTS.displayM.fontSize,
    lineHeight: FONTS.displayM.lineHeight,
  },
  extraLarge: {
    fontSize: FONTS.displayL.fontSize,
    lineHeight: FONTS.displayL.lineHeight,
  },
};

export const labelStyles = {
  false: {
    color: colorSelector('p400'),
    hover: {
      color: colorSelector('p400'),
    },
  },
  true: {
    backgroundColor: colorSelector('C100'),
    borderRadius: 20,
    hover: {
      color: colorSelector('C100'),
    },
  },
};

export const tabRootStyles = {
  default: {
    minHeight: '48px',
  },
  extraSmall: {
    minWidth: '80px',
    minHeight: '48px',
  },
  small: {
    minWidth: '100px',
    minHeight: '48px',
  },
  medium: {
    minWidth: '120px',
    minHeight: '48px',
  },
  large: {
    minWidth: '140px',

    minHeight: '48px',
  },
  extraLarge: {
    minWidth: '160px',

    minHeight: '48px',
  },
};

export const tabLabelContainerStyles = {
  default: {},
  extraSmall: {
    padding: '0 8px',
  },
  small: {
    padding: '0 15px',
  },
  medium: {},
  large: {},
  extraLarge: {},
};

export const rippleStyle = {
  color: colorSelector('P300'),
};
