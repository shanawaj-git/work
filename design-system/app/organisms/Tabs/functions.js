import { sizeStyles } from './styles';

const AVAILABLE_SIZES = Object.keys(sizeStyles);

export { AVAILABLE_SIZES };
