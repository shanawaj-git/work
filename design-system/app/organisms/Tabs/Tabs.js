import get from 'lodash.get';
import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';

import { AVAILABLE_SIZES } from './functions';
import { StyledTab, StyledTabs } from './styledComponents';

export const onChange = (setSelectedTab, onChangeCallback) => (e, value) => {
  e.preventDefault();
  setSelectedTab(value);

  if (onChangeCallback) {
    onChangeCallback(value);
  }
};

const Tabs = props => {
  const { preSelectedTab, tabs, onChangeCallback } = props;
  const initialSelectedTab = preSelectedTab || get(tabs, '[0].value');

  const [selectedTab, setSelectedTab] = useState(initialSelectedTab || 0);

  useEffect(() => {
    if (preSelectedTab) {
      setSelectedTab(preSelectedTab);
    }
  }, [preSelectedTab]);

  const renderTabs = () =>
    props.tabs.map((tab, index) => {
      const value = tab.value || index;
      const key = tab.id || `${props.id}-tab-${index}`;

      return (
        <StyledTab
          id={key}
          key={key}
          value={value}
          label={tab.label}
          size={props.size}
          data-testid={value}
          isActive={`${selectedTab === value}`}
        />
      );
    });

  const { id, ...rest } = props;

  return (
    <StyledTabs
      id={id}
      key={id}
      value={selectedTab}
      {...rest}
      onChange={onChange(setSelectedTab, onChangeCallback)}
    >
      {renderTabs()}
    </StyledTabs>
  );
};

Tabs.propTypes = {
  id: PropTypes.string,
  underline: PropTypes.bool,
  onChangeCallback: PropTypes.func,
  size: PropTypes.oneOf(AVAILABLE_SIZES),
  tabs: PropTypes.arrayOf(PropTypes.object),
  scrollButtons: PropTypes.oneOf(['auto', 'on', 'off']),
  variant: PropTypes.oneOf(['standard', 'scrollable', 'fullWidth']),
  preSelectedTab: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};

Tabs.defaultProps = {
  tabs: [],
  size: 'default',
  underline: false,
  variant: 'standard',
  scrollButtons: 'auto',
  preSelectedTab: null,
  onChangeCallback: null,
  id: `tabs-${Date.now()}`,
};

export default Tabs;
