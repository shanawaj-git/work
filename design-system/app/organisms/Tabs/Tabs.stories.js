import React from 'react';
import StoryWrapper from 'molecules/utility/StoryWrapper';
import Tabs from './Tabs';
import { AVAILABLE_SIZES } from './functions';

export default {
  title: 'Components/Containers/Tabs',
  component: Tabs,
  argTypes: {
    themeType: {
      control: { type: 'select', options: ['DEFAULT', 'MPC'] },
    },
    tabs: { table: { disable: true } },
    id: { table: { disable: true } },
    onChangeCallback: { table: { disable: true } },

    size: {
      control: { type: 'select', options: AVAILABLE_SIZES },
    },
  },
  decorators: [Story => <StoryWrapper>{Story()}</StoryWrapper>],
};

const defaultProps = {
  tabs: [{ label: 'delivery', value: 0 }, { label: 'tab & collect', value: 1 }],
};

const render = props => (
  <>
    <Tabs {...props} />
  </>
);

const Template = args => (
  <StoryWrapper themeVariant={args.themeType}>{render(args)}</StoryWrapper>
);

export const Standard = Template.bind({});
Standard.args = {
  ...defaultProps,
};

export const FullWidth = Template.bind({});
FullWidth.args = {
  ...defaultProps,
  variant: 'fullWidth',
};

export const scrollable = Template.bind({});
scrollable.args = {
  ...defaultProps,
  variant: 'scrollable',
  tabs: [
    ...defaultProps.tabs,
    { label: 'tab number #3', value: 2 },
    { label: 'tab number #4', value: 3 },
    { label: 'tab number #5', value: 4 },
    { label: 'tab number #6', value: 5 },
    { label: 'tab number #7', value: 6 },
    { label: 'tab number #8', value: 7 },
    { label: 'tab number #9', value: 8 },
    { label: 'tab number #10', value: 9 },
    { label: 'tab number #11', value: 10 },
    { label: 'tab number #12', value: 11 },
  ],
};

export const UnderlinedTabs = Template.bind({});
UnderlinedTabs.args = {
  ...defaultProps,
  variant: 'fullWidth',
  underline: true,
  size: 'extraSmall',
};

export const PreSelectedTab = Template.bind({});
PreSelectedTab.args = {
  ...defaultProps,
  preSelectedTab: 3,
  tabs: [
    ...defaultProps.tabs,
    { label: 'tab number #3', value: 2 },
    { label: 'tab number #4', value: 3 },
    { label: 'tab number #5', value: 4 },
  ],
};
