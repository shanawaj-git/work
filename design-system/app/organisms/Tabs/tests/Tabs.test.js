import React from 'react';
import renderer from 'react-test-renderer';
import Tabs, { onChange } from '../Tabs';

describe('onChange', () => {
  test('with onChangeCallback', () => {
    const setSelectedTab = jest.fn();
    const onChangeCallback = jest.fn();
    const event = { preventDefault: jest.fn() };

    onChange(setSelectedTab, onChangeCallback)(event, 'hello');

    expect(setSelectedTab).toHaveBeenCalledWith('hello');
    expect(onChangeCallback).toHaveBeenCalledWith('hello');
  });

  test('without onChangeCallback', () => {
    const setSelectedTab = jest.fn();
    const event = { preventDefault: jest.fn() };

    onChange(setSelectedTab)(event, 'hello');

    expect(setSelectedTab).toHaveBeenCalledWith('hello');
  });
});

describe('index', () => {
  const props = {
    id: 123,
  };
  it('Should import succesfully', () => {
    const tree = renderer.create(<Tabs {...props} />);
    expect(tree).toBeTruthy();
    expect(tree).toMatchSnapshot();
  });
});
