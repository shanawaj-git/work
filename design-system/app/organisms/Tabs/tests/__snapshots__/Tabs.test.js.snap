// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index Should import succesfully 1`] = `
<div
  className="root MuiTabs-root styledComponents__StyledTabs-sc-16g4axo-1 gllqKx css-1ujnqem-MuiTabs-root"
  id={123}
>
  <div
    className="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
    onScroll={[Function]}
    style={
      Object {
        "marginBottom": -0,
        "overflow": "hidden",
      }
    }
  >
    <div
      aria-orientation={null}
      className="flexContainer MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
      onKeyDown={[Function]}
      role="tablist"
    />
  </div>
</div>
`;
